File: 127-word-ladder/word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) 
    {
        unordered_set<string> set;
        for(string s : wordList)
        set.insert(s);

        queue<string> q;
        q.push(beginWord);
        set.erase(beginWord);

        int steps = 0;
        while(!q.empty())
        {
            int level = q.size();
            while(level--)
            {
                string word = q.front();
                q.pop();

                if(word == endWord)
                return steps + 1;

                for(int i = 0; i < word.length(); i++)
                {
                    char originalChar = word[i];
                    for(char c = 'a'; c <= 'z'; c++)
                    {
                        word[i] = c;
                        if(set.count(word) > 0)
                        {
                            set.erase(word);
                            q.push(word);
                        }
                    }
                    word[i] = originalChar;
                }
            }
            steps++;
        }
        return 0;
    }
};

File: 127-word-ladder/word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) 
    {
        Set<String> wordSet = new HashSet<>();
        for(String word : wordList)
        wordSet.add(word);

        if(!wordSet.contains(endWord))
        return 0;

        Queue<String> bfsQueue = new LinkedList<>();
        bfsQueue.add(beginWord);

        int steps = 1;
        while(!bfsQueue.isEmpty())
        {
            int level = bfsQueue.size();
            while(level > 0)
            {
                String currWord = bfsQueue.poll();

                if(currWord.equals(endWord))
                return steps;

                char[] wordArray = currWord.toCharArray();
                for(int i = 0; i < wordArray.length; i++)
                {
                    char originalCharacter = wordArray[i];
                    for(char ch = 'a'; ch <= 'z'; ch++)
                    {
                        wordArray[i] = ch;
                        String newWord = new String(wordArray);
                        if(wordSet.contains(newWord))
                        {
                            bfsQueue.add(newWord);
                            wordSet.remove(newWord);
                        }
                    }
                    // Restore original
                    wordArray[i] = originalCharacter; 
                }
                level--;
            }
            steps++;
        }
        return 0;
    }
}